pdf_tools.go

```diff
@@ -0,0 +1,180 @@
+package docconv
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	exts = []string{".jpg", ".tif", ".tiff", ".png", ".pbm"}
+)
+
+func compareExt(ext string, exts []string) bool {
+	for _, e := range exts {
+		if ext == e {
+			return true
+		}
+	}
+	return false
+}
+
+// Meta data
+type MetaResult struct {
+	meta map[string]string
+	err  error
+}
+
+type BodyResult struct {
+	body string
+	err  error
+}
+
+func ConvertImagePDF(path string) (BodyResult, error) {
+	bodyResult := BodyResult{}
+
+	tmp, err := ioutil.TempDir("/tmp", "tmp-imgs-")
+	if err != nil {
+		log.Println(err)
+		return bodyResult, err
+	}
+	tmpDir := fmt.Sprintf("%s/", tmp)
+	defer os.RemoveAll(tmpDir)
+
+	_, err = exec.Command("pdfimages", "-j", path, tmpDir).Output()
+	if err != nil {
+		return bodyResult, err
+	}
+
+	files := []string{}
+
+	walkFunc := func(path string, info os.FileInfo, err error) error {
+		path, err = filepath.Abs(path)
+		if err != nil {
+			return err
+		}
+
+		if compareExt(filepath.Ext(path), exts) {
+			files = append(files, path)
+		}
+		return nil
+	}
+	filepath.Walk(tmpDir, walkFunc)
+
+	var wg sync.WaitGroup
+
+	m := struct {
+		sync.Mutex
+		fileMap map[int]string
+	}{}
+
+	m.fileMap = make(map[int]string)
+
+	wg.Add(len(files))
+	for indx, p := range files {
+		go func(idx int, pathFile string, ww *sync.WaitGroup) {
+			defer ww.Done()
+			f, err := os.Open(pathFile)
+			if err != nil {
+				log.Println(err)
+			}
+			out, _, err := ConvertImage(f)
+			if err != nil {
+				bodyResult.err = err
+			}
+
+			m.Lock()
+			m.fileMap[idx] = out
+			m.Unlock()
+
+			f.Close()
+		}(indx, p, &wg)
+	}
+	wg.Wait()
+
+	o := make([]string, len(m.fileMap))
+
+	for i := 0; i < len(m.fileMap); i++ {
+		o = append(o, m.fileMap[i])
```

Comment:
No need to create an intermediate slice here. Note: the fileMap itself could also just be a channel to send the body back on, which would also remove the need for the mutex.

```go
for _, str := range m.fileMap {
    bodyResult += str + " " 
}
```